File: src/Aeson_encode.ml
type 'a encoder = 'a -> Js.Json.t

external null : Js.Json.t = "null" [@@bs.val]
external string : string -> Js.Json.t = "%identity"
external float : float -> Js.Json.t = "%identity"
external int : int -> Js.Json.t = "%identity"
external int32 : int32 -> Js.Json.t = "%identity"
external int64_to_array : int64 -> Js.Json.t = "%identity"
external nativeint : nativeint -> Js.Json.t = "%identity"
external bool : bool -> Js.Json.t = "%identity" 
external dict : Js.Json.t Js_dict.t -> Js.Json.t = "%identity"
                                                 
let int64_to_string (x: Int64.t) = string (Int64.to_string x)
                                                 
let bigint (x: Bigint.t) = string (Bigint.to_string x)
                                                 
let uint8 (x: U.UInt8.t) = int (U.UInt8.toInt x)

let uint16 (x: U.UInt16.t) = int (U.UInt16.toInt x)

(* underlying it is an int64 but is less than JS limit,
   haskell expects a numeric literal *)                           
let uint32 (x: U.UInt32.t) = int (U.UInt32.toInt x)

let uint64 (x: U.UInt64.t) = string (U.UInt64.toString x)
  
let nullable encode = function
  | None -> null
  | Some v -> encode v

let withDefault d encode = function
  | None -> d
  | Some v -> encode v

let optional encode optionalValue =
  match optionalValue with
  | Some value -> encode value
  | None -> null

let optionalField fieldName encode optionalValue =
  match optionalValue with
  | Some value -> [(fieldName, encode value)]
  | None -> []

(* Haskell aeson renders .000Z as Z *)          
let date d: Js.Json.t = string (Js.String.replace ".000Z" "Z" (Js_date.toISOString d))
  
let object_ props: Js.Json.t =
  props |> Js.Dict.fromList
        |> dict

external array : Js.Json.t array -> Js.Json.t = "%identity"

let list encode l =
  l |> List.map encode
    |> Array.of_list
    |> array

let pair encodeT0 encodeT1 tuple =
  let (t0, t1) = tuple in
  array [| encodeT0 t0 ; encodeT1 t1 |]

let tuple2 = pair

let beltMap encodeKey encodeValue obj =
  list (pair encodeKey encodeValue) (Array.to_list (Belt.Map.toArray obj))

let beltMapInt encodeValue obj =
  object_
  (List.map (fun (k, v) -> (string_of_int k, encodeValue v)) (Array.to_list (Belt.Map.Int.toArray obj)))

let beltMapString encodeValue obj =
  object_
  (List.map (fun (k, v) -> (k, encodeValue v)) (Array.to_list (Belt.Map.String.toArray obj)))

let tuple3 encodeT0 encodeT1 encodeT2 tuple =
  let (t0, t1, t2) = tuple in
  array [| encodeT0 t0 ; encodeT1 t1 ; encodeT2 t2 |]

let tuple4 encodeT0 encodeT1 encodeT2 encodeT3 tuple =
  let (t0, t1, t2, t3) = tuple in
  array [| encodeT0 t0 ; encodeT1 t1 ; encodeT2 t2 ; encodeT3 t3 |]

let tuple5 encodeT0 encodeT1 encodeT2 encodeT3 encodeT4 tuple =
  let (t0, t1, t2, t3, t4) = tuple in
  array [| encodeT0 t0 ; encodeT1 t1 ; encodeT2 t2 ; encodeT3 t3 ; encodeT4 t4 |]

let tuple6 encodeT0 encodeT1 encodeT2 encodeT3 encodeT4 encodeT5 tuple =
  let (t0, t1, t2, t3, t4, t5) = tuple in
  array [| encodeT0 t0 ; encodeT1 t1 ; encodeT2 t2 ; encodeT3 t3 ; encodeT4 t4 ; encodeT5 t5 |]

let tuple7 encodeT0 encodeT1 encodeT2 encodeT3 encodeT4 encodeT5 encodeT6 tuple =
  let (t0, t1, t2, t3, t4, t5, t6) = tuple in
  array [| encodeT0 t0 ; encodeT1 t1 ; encodeT2 t2 ; encodeT3 t3 ; encodeT4 t4 ; encodeT5 t5 ; encodeT6 t6 |]

let tuple8 encodeT0 encodeT1 encodeT2 encodeT3 encodeT4 encodeT5 encodeT6 encodeT7 tuple =
  let (t0, t1, t2, t3, t4, t5, t6, t7) = tuple in
  array [| encodeT0 t0 ; encodeT1 t1 ; encodeT2 t2 ; encodeT3 t3 ; encodeT4 t4 ; encodeT5 t5 ; encodeT6 t6 ; encodeT7 t7 |]

let tuple9 encodeT0 encodeT1 encodeT2 encodeT3 encodeT4 encodeT5 encodeT6 encodeT7 encodeT8 tuple =
  let (t0, t1, t2, t3, t4, t5, t6, t7, t8) = tuple in
  array [| encodeT0 t0 ; encodeT1 t1 ; encodeT2 t2 ; encodeT3 t3 ; encodeT4 t4 ; encodeT5 t5 ; encodeT6 t6 ; encodeT7 t7 ; encodeT8 t8 |]

let tuple10 encodeT0 encodeT1 encodeT2 encodeT3 encodeT4 encodeT5 encodeT6 encodeT7 encodeT8 encodeT9 tuple =
  let (t0, t1, t2, t3, t4, t5, t6, t7, t8, t9) = tuple in
  array [| encodeT0 t0 ; encodeT1 t1 ; encodeT2 t2 ; encodeT3 t3 ; encodeT4 t4 ; encodeT5 t5 ; encodeT6 t6 ; encodeT7 t7 ; encodeT8 t8 ; encodeT9 t9 |]

let result encodeA encodeB e =
  match e with
  | Belt.Result.Ok a -> object_ [("Ok", encodeA a)]
  | Belt.Result.Error b -> object_ [("Error", encodeB b)]

let either encodeL encodeR e =
  match e with
  | Aeson_compatibility.Either.Left l -> object_ [("Left", encodeL l)]
  | Aeson_compatibility.Either.Right r -> object_ [("Right", encodeR r)]

let singleEnumerator _x =  array [| |]

external stringArray : string array -> Js.Json.t = "%identity"
external numberArray : float array -> Js.Json.t = "%identity"
external boolArray : bool array -> Js.Json.t = "%identity"

File: lib/js/src/Aeson_encode.js
'use strict';

var Js_dict = require("rescript/lib/js/js_dict.js");
var Js_json = require("rescript/lib/js/js_json.js");
var Belt_Map = require("rescript/lib/js/belt_Map.js");
var Js_string = require("rescript/lib/js/js_string.js");
var Core__List = require("@rescript/core/lib/js/src/Core__List.js");
var Belt_MapInt = require("rescript/lib/js/belt_MapInt.js");
var Caml_option = require("rescript/lib/js/caml_option.js");
var Belt_MapString = require("rescript/lib/js/belt_MapString.js");

function bigint(x) {
  return x.toString();
}

function nullable(encode, x) {
  if (x !== undefined) {
    return encode(Caml_option.valFromOption(x));
  } else {
    return null;
  }
}

function withDefault(d, encode, x) {
  if (x !== undefined) {
    return encode(Caml_option.valFromOption(x));
  } else {
    return d;
  }
}

function optional(encode, optionalValue) {
  if (optionalValue !== undefined) {
    return encode(Caml_option.valFromOption(optionalValue));
  } else {
    return null;
  }
}

function optionalField(fieldName, encode, optionalValue) {
  if (optionalValue !== undefined) {
    return {
            hd: [
              fieldName,
              encode(Caml_option.valFromOption(optionalValue))
            ],
            tl: /* [] */0
          };
  } else {
    return /* [] */0;
  }
}

function date(d) {
  return Js_string.replace(".000Z", "Z", d.toISOString());
}

var object_ = Js_dict.fromList;

function list(encode, l) {
  return Core__List.toArray(Core__List.map(l, (function (x) {
                    return encode(x);
                  })));
}

function pair(encodeT0, encodeT1, tuple) {
  return [
          encodeT0(tuple[0]),
          encodeT1(tuple[1])
        ];
}

function beltMap(encodeKey, encodeValue, obj) {
  return list((function (extra) {
                return pair(encodeKey, encodeValue, extra);
              }), Core__List.fromArray(Belt_Map.toArray(obj)));
}

function beltMap1(encodeKey, encodeValue, obj) {
  var xs = Belt_Map.toArray(obj);
  var encodeKey1 = function (key) {
    var str = Js_json.classify(encodeKey(key));
    if (typeof str !== "object" || str.TAG !== "JSONString") {
      return JSON.stringify(encodeKey(key));
    } else {
      return str._0;
    }
  };
  var xs$1 = xs.map(function (param) {
        return [
                encodeKey1(param[0]),
                encodeValue(param[1])
              ];
      });
  return Js_dict.fromList(Core__List.fromArray(xs$1));
}

function beltMapInt(encodeValue, obj) {
  return Js_dict.fromList(Core__List.map(Core__List.fromArray(Belt_MapInt.toArray(obj)), (function (param) {
                    return [
                            String(param[0]),
                            encodeValue(param[1])
                          ];
                  })));
}

function beltMapString(encodeValue, obj) {
  return Js_dict.fromList(Core__List.map(Core__List.fromArray(Belt_MapString.toArray(obj)), (function (param) {
                    return [
                            param[0],
                            encodeValue(param[1])
                          ];
                  })));
}

function tuple3(encodeT0, encodeT1, encodeT2, tuple) {
  return [
          encodeT0(tuple[0]),
          encodeT1(tuple[1]),
          encodeT2(tuple[2])
        ];
}

function tuple4(encodeT0, encodeT1, encodeT2, encodeT3, tuple) {
  return [
          encodeT0(tuple[0]),
          encodeT1(tuple[1]),
          encodeT2(tuple[2]),
          encodeT3(tuple[3])
        ];
}

function tuple5(encodeT0, encodeT1, encodeT2, encodeT3, encodeT4, tuple) {
  return [
          encodeT0(tuple[0]),
          encodeT1(tuple[1]),
          encodeT2(tuple[2]),
          encodeT3(tuple[3]),
          encodeT4(tuple[4])
        ];
}

function tuple6(encodeT0, encodeT1, encodeT2, encodeT3, encodeT4, encodeT5, tuple) {
  return [
          encodeT0(tuple[0]),
          encodeT1(tuple[1]),
          encodeT2(tuple[2]),
          encodeT3(tuple[3]),
          encodeT4(tuple[4]),
          encodeT5(tuple[5])
        ];
}

function tuple7(encodeT0, encodeT1, encodeT2, encodeT3, encodeT4, encodeT5, encodeT6, tuple) {
  return [
          encodeT0(tuple[0]),
          encodeT1(tuple[1]),
          encodeT2(tuple[2]),
          encodeT3(tuple[3]),
          encodeT4(tuple[4]),
          encodeT5(tuple[5]),
          encodeT6(tuple[6])
        ];
}

function tuple8(encodeT0, encodeT1, encodeT2, encodeT3, encodeT4, encodeT5, encodeT6, encodeT7, tuple) {
  return [
          encodeT0(tuple[0]),
          encodeT1(tuple[1]),
          encodeT2(tuple[2]),
          encodeT3(tuple[3]),
          encodeT4(tuple[4]),
          encodeT5(tuple[5]),
          encodeT6(tuple[6]),
          encodeT7(tuple[7])
        ];
}

function tuple9(encodeT0, encodeT1, encodeT2, encodeT3, encodeT4, encodeT5, encodeT6, encodeT7, encodeT8, tuple) {
  return [
          encodeT0(tuple[0]),
          encodeT1(tuple[1]),
          encodeT2(tuple[2]),
          encodeT3(tuple[3]),
          encodeT4(tuple[4]),
          encodeT5(tuple[5]),
          encodeT6(tuple[6]),
          encodeT7(tuple[7]),
          encodeT8(tuple[8])
        ];
}

function tuple10(encodeT0, encodeT1, encodeT2, encodeT3, encodeT4, encodeT5, encodeT6, encodeT7, encodeT8, encodeT9, tuple) {
  return [
          encodeT0(tuple[0]),
          encodeT1(tuple[1]),
          encodeT2(tuple[2]),
          encodeT3(tuple[3]),
          encodeT4(tuple[4]),
          encodeT5(tuple[5]),
          encodeT6(tuple[6]),
          encodeT7(tuple[7]),
          encodeT8(tuple[8]),
          encodeT9(tuple[9])
        ];
}

function result(encodeA, encodeB, e) {
  if (e.TAG === "Ok") {
    return Js_dict.fromList({
                hd: [
                  "Ok",
                  encodeA(e._0)
                ],
                tl: /* [] */0
              });
  } else {
    return Js_dict.fromList({
                hd: [
                  "Error",
                  encodeB(e._0)
                ],
                tl: /* [] */0
              });
  }
}

function either(encodeL, encodeR, e) {
  if (e.TAG === "Left") {
    return Js_dict.fromList({
                hd: [
                  "Left",
                  encodeL(e._0)
                ],
                tl: /* [] */0
              });
  } else {
    return Js_dict.fromList({
                hd: [
                  "Right",
                  encodeR(e._0)
                ],
                tl: /* [] */0
              });
  }
}

function singleEnumerator(_x) {
  return [];
}

var tuple2 = pair;

exports.bigint = bigint;
exports.nullable = nullable;
exports.withDefault = withDefault;
exports.object_ = object_;
exports.optional = optional;
exports.optionalField = optionalField;
exports.date = date;
exports.beltMap = beltMap;
exports.beltMap1 = beltMap1;
exports.beltMapInt = beltMapInt;
exports.beltMapString = beltMapString;
exports.list = list;
exports.pair = pair;
exports.tuple2 = tuple2;
exports.tuple3 = tuple3;
exports.tuple4 = tuple4;
exports.tuple5 = tuple5;
exports.tuple6 = tuple6;
exports.tuple7 = tuple7;
exports.tuple8 = tuple8;
exports.tuple9 = tuple9;
exports.tuple10 = tuple10;
exports.result = result;
exports.either = either;
exports.singleEnumerator = singleEnumerator;
/* No side effect */
